### PR TITLE
[Text Generation] Support for causal masks, internal KV cache, and initial testing framework

### PR DESCRIPTION
**Pull request by dbogunowicz**
1. Making sure that models that have not been reexported do not raise errors.
2. Verbose messages for the user regarding the enablement/disablement of multitoken engine for prompt processing.

# INCLUDED PRs:
Includes PR #1163 which adds testing and internal KV cache support, made up of: 

- [x] - #1154 
- [x] - #1155
- [x] - #1162 
- [x] - #1135

### CODE REVIEW

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 1]
```diff
@@ -180,24 +183,49 @@ def __init__(
...
-        )
+        self.engine, self.multitoken_engine = None, None
+
+        """
+        Inititalize multitoken engine only if:
+         - model has kv cache support and multitoken prefill is enabled
+         or
+         - model does not have kv cache support
+
+         Initialize single token engine only if model has kv cache support
```
Comment (dbogunowicz) [inline src/deepsparse/transformers/pipelines/text_generation.py:194]:
```suggestion
        Initialize single token engine only if model has kv cache support
```

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 2]
```diff
@@ -180,24 +183,49 @@ def __init__(
...
-            deterministic=self.deterministic,
-            sequence_length=self.sequence_length,
-            input_ids_length=prompt_processing_sequence_length,
-            tokenizer=self.tokenizer,
-            use_deepsparse_cache=use_deepsparse_cache,
-        )
+        self.engine, self.multitoken_engine = None, None
+
+        """
+        Inititalize multitoken engine only if:
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:189]: additionally, __init__ is getting quite long, can we break some out into helper functions?

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 3]
```diff
@@ -180,24 +183,49 @@ def __init__(
...
+         or
+         - model does not have kv cache support
+
+        Initialize single token engine only if model has kv cache support
+        """
+        if self.cache_support_enabled:
+            if not self.enable_multitoken_prefill:
+                warnings.warn(
+                    "The ONNX graph does not support processing the prompt in "
+                    "multitoken fashion. The prompt will be processed in "
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:200]: let's choose different words instead of "fashion", also instead of multitoken and autoregressive, let's use sequence length greater than 1 etc since this is more understandable externally

File: src/deepsparse/utils/onnx.py [hunk 4]
```diff
@@ -50,10 +50,13 @@
     "truncate_onnx_model",
     "truncate_onnx_embedding_model",
     "default_cached_outputs",
+    "CACHE_OUTPUT_NAME",
 ]
 
 _LOGGER = logging.getLogger(__name__)
 
+CACHE_OUTPUT_NAME = "present"
```
Comment (bfineran) [inline src/deepsparse/utils/onnx.py:58]:
```suggestion
CACHE_OUTPUT_PREFIX = "present"
```

File: src/deepsparse/utils/onnx.py [hunk 5]
```diff
@@ -475,20 +478,19 @@ def truncate_onnx_embedding_model(
...
 
-    # Create a boolean list of every output of the
-    # model [logits, key0, value0, key1, value1, ..., keyN, valueN]
-    cached_outputs = [True for i in range(len(outputs))]
-
-    # Assume first input is logits and logits ought not to be cached
-    cached_outputs[0] = False
-
-    return cached_outputs
+    return [output.name.startswith(CACHE_OUTPUT_NAME) for output in outputs]
```
Comment (bfineran) [inline src/deepsparse/utils/onnx.py:496]:
```suggestion
    return [output.name.startswith(CACHE_OUTPUT_PREFIX) for output in outputs]
```

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 6]
```diff
@@ -164,62 +166,110 @@ def __init__(
...
+
+        if self.cache_support_enabled:
+            # emit the appropriate user message depending whether we are
+            # instantiation the multitoken engine or not
+            if not self.enable_multitoken_prefill:
+                warnings.warn(
+                    "The ONNX graph does not support processing the prompt in "
+                    "a single pass. The creation of an auxiliary engine for "
+                    "processing the prompt efficiently is disabled. "
+                    "The prompt will be processed in autoregressive fashion."
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:230]:
```suggestion
                    "This ONNX graph does not support processing the prompt in "
                    "with processing length > 1. Creation of an auxiliary engine for "
                    "processing the prompt at a larger seq_len is disabled. "
                    "The prompt will be processed in with processing length 1."
```

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 7]
```diff
@@ -164,62 +166,110 @@ def __init__(
...
+                )
+            else:
+                _LOGGER.info(
+                    "Creating an auxiliary engine to process a prompt in a "
+                    "a single pass. This guarantees better performance, but "
+                    "may result in additional memory consumption."
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:236]:
```suggestion
                    "Compiling an auxiliary engine to process a prompt with a "
                    "larger processing length. This improves performance, but "
                    "may result in additional memory consumption."
```

File: src/deepsparse/transformers/pipelines/text_generation.py [hunk 8]
```diff
@@ -581,13 +638,13 @@ def engine_inputs_for_prefill(
...
 
     @property
-    def has_cache(self) -> bool:
+    def cache_support_enabled(self) -> bool:
         """
         Returns whether the ran model has kv cache or not
 
         :return: True if the model has kv cache, False otherwise
         """
-        return self.multitoken_engine.kv_cache_enabled
```
Comment (bfineran) [inline src/deepsparse/transformers/pipelines/text_generation.py:590]: store as private variable that is initialized to `None` in `__init__` and filled out on first call here, referenced after

**Review by bfineran — DISMISSED** [review]

**Review by Satrat — DISMISSED** [review]

**Review by Satrat — DISMISSED** [review]

**Review by bfineran — APPROVED** [review]

**Review by bfineran — APPROVED** [review]
Re-approving - includes changes from #1163
